pre-docs/writing_yaml_plans.md

```diff
@@ -0,0 +1,294 @@
+
+`steps`: The list of steps to perform
+`parameters`: The parameters accepted by the plan (optional)
+`return`: The value to return from the plan (optional)
+
+## Parameters
+
+A plan can accept a set of parameters, specified in the `parameters` key. The value of `parameters` is a map, where each key is the name of a parameter and the value is a map describing the parameter. A parameter can have a `type`, a `default`, and a `description`.
+
+For example, this plan accepts a `load_balancer` name as a string, and two sets of nodes called `frontends` and `backends`. It then runs a series of tasks to remove nodes from the load balancer, update the application, and then add them back to the load balancer.
+
+```yaml
+parameters:
+  load_balancer:
+    type: String
+    description: "The application load balancer"
+  frontends:
+    type: TargetSpec
+    description: "The list of application frontend servers"
+  version:
+    type: String
+    description: "The new application version to deploy"
+
+steps:
+  - task: mymodule::lb_remove
+    target: $load_balancer
+    parameters:
+      frontends: $frontends
+  - task: mymodule::update_frontend_app
+    target: $frontends
+    parameters:
+      version: $version
+  - task: mymodule::lb_add
+    target: $load_balancer
+    parameters:
+      frontends: $frontends
+```
+
+The `type` must be a valid Puppet datatype.
+
+Parameter values can be referenced from steps as Puppet variables.
+
+## Plan steps
+
+The `steps` key is an array of step objects, each of which corresponds to a specific action to take.
+
+When the plan is run, each step will be executed in order. If the step has a `name` field, the result of the step will be stored in a variable with the name of the step and be referenced from later steps.
+
+If a step fails, the plan will halt execution and return the result of the failed step.
+
+Steps can also have a `description` field to explain what the step is doing.
+
+The other available keys depend on the kind of step. The available steps are task, command, script, file upload, plan, and eval.
+
+### Task step
+
+A task step will run a task on a list of targets and return the results.
+
+Fields:
+`task`: Which task to run
+`target`: A target or list of targets to run the task on
+`parameters`: A map of parameter values to pass to the task (optional)
+
+Example:
+
+```yaml
+steps:
+  - task: package
+    target: 'webservers'
+    description: "Check the version of the openssl package on the webservers"
+    parameters:
+      action: status
+      name: openssl
+```
+
+### Command step
+
+A command step will run a single command on a list of targets and return the results, containing stdout, stderr and exit code.
+
+Fields:
+`command`: Which command to run
+`target`: A target or list of targets to run the command on
+
+Example:
+
+```yaml
+steps:
+  - command: hostname -f
+    target: 'webservers'
+    description: "Get the webserver hostnames"
```

Comment:
As a matter of style I should description go first? I guess this order most closely maps to the functions.

Reply:
I personally find it easier to scan what the step is doing by looking at the command/task and target, so I put those first. Not sure which makes more sense.